lib/pigeon/gcm_worker.ex

```diff
@@ -0,0 +1,244 @@
+    new_stream_id = stream_id + 2
+    { :noreply, %{state | stream_id: new_stream_id, queue: new_q } }
+  end
+
+  defp parse_error(data) do
+    {:ok, response} = Poison.decode(data)
+    response["reason"] |> Macro.underscore |> String.to_existing_atom
+  end
+
+  defp log_error(code, reason) do
+    Logger.error("#{reason}: #{code}")
+  end
+
+
+  def handle_info(:ping, state) do
+    Kadabra.ping(state.gcm_socket)
+    Process.send_after(self(), :ping, @ping_period)
+
+    { :noreply, state }
+  end
+
+  def handle_info({:end_stream, %Kadabra.Stream{id: stream_id, headers: headers, body: body}},
+                                                %{gcm_socket: _socket, queue: queue} = state) do
+
+    {registration_ids, on_response} = queue["#{stream_id}"]
+    Logger.debug("#{inspect body} and #{inspect headers}")
+    case get_status(headers) do
+      "200" ->
+        result =  Poison.decode! body
+        parse_result(registration_ids, result, on_response)
+        new_queue = Map.delete(queue, "#{stream_id}")
+        {:noreply, %{state | queue: new_queue}}
+      nil ->
+        {:noreply, state}
+      "401" ->
+        log_error("401", "Unauthorized")
+        unless on_response == nil do on_response.({:error, :unauthorized}) end
+        new_queue = Map.delete(queue, "#{stream_id}")
+        {:noreply, %{state | queue: new_queue}}
+      "400" ->
+        log_error("400", "Malformed JSON")
+        unless on_response == nil do on_response.({:error, :malformed_json}) end
+        new_queue = Map.delete(queue, "#{stream_id}")
+        {:noreply, %{state | queue: new_queue}}
+      code ->
+        reason = parse_error(body)
+        log_error(code, reason)
+        unless on_response == nil do on_response.({:error, reason}) end
+        new_queue = Map.delete(queue, "#{stream_id}")
+        {:noreply, %{state | queue: new_queue}}
+    end
+  end
+
+  def handle_info({:ping, _from}, state), do: {:noreply, state}
+
+  def handle_info({:closed, _from}, state), do: {:noreply, state}
+
+  def handle_info({:ok, _from}, state), do: {:noreply, state}
+
+  # no on_response callback, ignore
+  def parse_result(_, _, nil), do: :ok
+
+  def parse_result(ids, %{"results" => results}, on_response) do
+    parse_result1(ids, results, on_response, %NotificationResponse{})
+  end
+
+  def parse_result1([], [], on_response, result) do
+    on_response.({:ok, result})
+  end
+
+  def parse_result1(regid, results, on_response, result) when is_binary(regid) do
+    parse_result1([regid], results, on_response, result)
+  end
+
+  def parse_result1([regid | reg_res],
+                    [%{"message_id" => id, "registration_id" => new_regid} | rest_results],
+                    on_response,
+                    %NotificationResponse{ update: update} =  resp) do
+
+    new_updates = [{regid, new_regid} | update]
+    parse_result1(reg_res, rest_results, on_response, %{resp | message_id: id, update: new_updates})
```

Comment:
I'm not sure. [Documentation](https://firebase.google.com/docs/cloud-messaging/http-server-ref) states table 5:
`message_id: String specifying a unique ID for each successfully processed message.`

Given the location of the information, the fact that multicast id exists, I'd say let's just pass it along in the `NotificationResponse.result` array and set message_id to multicast_id (and rename it to multicast_id)